amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCleanupManager.java

```diff
@@ -0,0 +1,299 @@
+import software.amazon.kinesis.leases.exceptions.InvalidStateException;
+import software.amazon.kinesis.leases.exceptions.ProvisionedThroughputException;
+import software.amazon.kinesis.metrics.MetricsFactory;
+import software.amazon.kinesis.retrieval.AWSExceptionManager;
+import software.amazon.kinesis.retrieval.kpl.ExtendedSequenceNumber;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to cleanup of any expired/closed shard leases. It will cleanup leases periodically as defined by
+ * {@link LeaseManagementConfig#leaseCleanupIntervalMillis()} asynchronously.
+ */
+@Accessors(fluent=true)
+@Slf4j
+@RequiredArgsConstructor
+@EqualsAndHashCode
+public class LeaseCleanupManager {
+    @NonNull
+    private final LeaseCoordinator leaseCoordinator;
+    @NonNull
+    private final KinesisAsyncClient kinesisClient;
+    @NonNull
+    private final MetricsFactory metricsFactory;
+    @NonNull
+    private final Duration maxFutureWait;
+    @NonNull
+    private final ScheduledExecutorService deletionThreadPool;
+    private final boolean cleanupLeasesUponShardCompletion;
+    private final long leaseCleanupIntervalMillis;
+    private final long completedLeaseCleanupThresholdMillis;
+    private final long garbageLeaseCleanupThresholdMillis;
+
+    private final Queue<LeasePendingDeletion> deletionQueue = new ConcurrentLinkedQueue<>();
+
+    private static final int MAX_RECORDS = 1;
+    private static final long INITIAL_DELAY = 0L;
+    private static final String LEASE_CLEANUP_MANAGER = "LeaseCleanupManager";
+
+    @Getter
+    private volatile boolean isRunning = false;
+
+    /**
+     * Starts the lease cleanup thread, which is scheduled periodically as specified by
+     * {@link LeaseCleanupManager#leaseCleanupIntervalMillis}
+     */
+    public void start() {
+        log.debug("Starting lease cleanup thread.");
+        deletionThreadPool.scheduleAtFixedRate(new LeaseCleanupThread(), INITIAL_DELAY, leaseCleanupIntervalMillis,
+                TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Enqueues a lease for deletion.
+     * @param leasePendingDeletion
+     */
+    public void enqueueForDeletion(LeasePendingDeletion leasePendingDeletion) {
+        final Lease lease = leasePendingDeletion.lease();
+        if (lease == null) {
+            log.warn("Cannot enqueue lease {} for deferred deletion - instance doesn't hold the lease for that shard.",
+                    lease.leaseKey());
+        } else {
+            if (!deletionQueue.contains(leasePendingDeletion)) {
```

Comment:
Hmmm, this contains check complexity is O(n). This will not be performant especially when we have large number of leases to be processed. Can we move to LinkedHashMap in next iteration. Create a SIM and add it here in comments

Reply:
Added a TODO